### PR TITLE
HDDS-3926. OM Token Identifier table should use in-house serialization.

### PR DESCRIPTION
**Pull request by prashantpogde**
## What changes were proposed in this pull request?

Ozone Manager Token Identifier table should use in-house serialization rather than rely on proto serialization for key.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3926

## How was this patch tested?

CleanBuild and Integration Tests

### CODE REVIEW

File: hadoop-hdds/interface-client/src/main/proto/proto.lock [hunk 1]
```diff
@@ -1476,6 +1476,21 @@
           }
         ],
         "messages": [
+          {
```
Comment (xiaoyuyao) [inline hadoop-hdds/interface-client/src/main/proto/proto.lock:1479]: The proto change seems unrelated to this PR.

Comment (prashantpogde) [reply hadoop-hdds/interface-client/src/main/proto/proto.lock:1479]: It was auto generated. I did not modify anything here in this file.

Comment (prashantpogde) [reply hadoop-hdds/interface-client/src/main/proto/proto.lock:1479]: Integration test failure seems unrelated to the changes here.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 2]
```diff
@@ -77,6 +79,119 @@ public Text getKind() {
     return KIND_NAME;
   }
 
+  /** Instead of relying on proto serialization, this
+   *  provides  explicit serialization for OzoneTokenIdentifier.
+   * @return byte[]
+   */
+  public byte[] toUniqueSerilizedKey() {
```
Comment (adoroszlai) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:86]:
```suggestion
  public byte[] toUniqueSerializedKey() {
```

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 3]
```diff
@@ -77,6 +79,119 @@ public Text getKind() {
...
+          "Can't encode the the raw data ", e);
+    }
+    return result.array();
+  }
+
+  /** Instead of relying on proto deserialization, this
+   *  provides  explicit deserialization for OzoneTokenIdentifier.
+   * @return byte[]
+   */
+  public static OzoneTokenIdentifier fromUniqueSerilizedKey(byte[] rawData) {
```
Comment (adoroszlai) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:138]:
```suggestion
  public static OzoneTokenIdentifier fromUniqueSerializedKey(byte[] rawData) {
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java [hunk 4]
```diff
@@ -33,7 +33,7 @@
   public byte[] toPersistedFormat(OzoneTokenIdentifier object) {
     Preconditions
         .checkNotNull(object, "Null object can't be converted to byte array.");
-    return object.getBytes();
+    return object.toUniqueSerilizedKey();
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java:36]:
```suggestion
    return object.toUniqueSerializedKey();
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java [hunk 5]
```diff
@@ -42,8 +42,8 @@ public OzoneTokenIdentifier fromPersistedFormat(byte[] rawData)
     Preconditions.checkNotNull(rawData,
         "Null byte array can't converted to real object.");
     try {
-      return OzoneTokenIdentifier.readProtoBuf(rawData);
-    } catch (InvalidProtocolBufferException e) {
+      return OzoneTokenIdentifier.fromUniqueSerilizedKey(rawData);
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java:45]:
```suggestion
      return OzoneTokenIdentifier.fromUniqueSerializedKey(rawData);
```

File: hadoop-hdds/interface-client/src/main/proto/proto.lock (hunk 1 above)
Comment (adoroszlai) [reply hadoop-hdds/interface-client/src/main/proto/proto.lock:1479]: > It was auto generated. I did not modify anything here in this file.

Please exclude both proto.lock files from the patch.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 6]
```diff
@@ -77,6 +79,119 @@ public Text getKind() {
...
+    result.order(ByteOrder.BIG_ENDIAN);
+    try {
+      result.putLong(getIssueDate());
+      result.putInt(getMasterKeyId());
+      result.putInt(getSequenceNumber());
+
+      result.putLong(getMaxDate());
+
+      result.putInt(getOwner().toString().length());
+      result.put(getOwner().toString().getBytes());
```
Comment (adoroszlai) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:98]: Findbugs:

```
H I Dm: Found reliance on default encoding in org.apache.hadoop.ozone.security.OzoneTokenIdentifier.toUniqueSerilizedKey(): String.getBytes()  At OzoneTokenIdentifier.java:[line 98]
```

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 7]
```diff
@@ -77,6 +79,119 @@ public Text getKind() {
...
+    byte[] renewerBytes = new byte[strsize];
+    inbuf.get(renewerBytes);
+    result.setRenewer(new Text(renewerBytes));
+
+    // Set s3 specific fields.
+    if (inbuf.getInt() == S3AUTHINFO.getNumber()) {
+      strsize = inbuf.getInt();
+      byte[] awsAccessIdBytes = new byte[strsize];
+      inbuf.get(awsAccessIdBytes);
+      result.setAwsAccessId(new String(awsAccessIdBytes));
```
Comment (adoroszlai) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:169]: Findbugs:

```
H I Dm: Found reliance on default encoding in org.apache.hadoop.ozone.security.OzoneTokenIdentifier.fromUniqueSerilizedKey(byte[]): new String(byte[])  At OzoneTokenIdentifier.java:[line 169]
```

File: hadoop-hdds/interface-client/src/main/proto/proto.lock (hunk 1 above)
Comment (prashantpogde) [reply hadoop-hdds/interface-client/src/main/proto/proto.lock]: > > It was auto generated. I did not modify anything here in this file.
> 
> Please exclude both proto.lock files from the patch.

done

Comment (prashantpogde) [reply hadoop-hdds/interface-client/src/main/proto/proto.lock]: > The proto.lock change seems unrelated to this PR.

these files are reverted.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java (hunk 2 above)
Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java]: done

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java (hunk 3 above)
Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java]: done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java (hunk 4 above)
Comment (prashantpogde) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java]: done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java (hunk 5 above)
Comment (prashantpogde) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java]: done

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java (hunk 6 above)
Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java]: done

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java (hunk 7 above)
Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java]: done

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 8]
```diff
@@ -77,6 +80,122 @@ public Text getKind() {
     return KIND_NAME;
   }
 
+  /** Instead of relying on proto serialization, this
+   *  provides  explicit serialization for OzoneTokenIdentifier.
+   * @return byte[]
+   */
+  public byte[] toUniqueSerializedKey() {
+    ByteBuffer result =
+        ByteBuffer.allocate(4096);
```
Comment (ChenSammi) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:89]: I see there is StrToSign if it's S3 auth. I 'm curious about how big this StrToSign can be? And whether 4096 bytes can hold all these.

Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:89]: > I see there is StrToSign if it's S3 auth. I 'm curious about how big this StrToSign can be? And whether 4096 bytes can hold all these.

It should not be too big and not cause overflow.

https://docs.aws.amazon.com/general/latest/gr/sigv4-create-string-to-sign.html

I saw something similar in other places too e.g. 
org/apache/hadoop/security/token/TokenIdentifier.class with 4096 bytes limit.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 9]
```diff
@@ -77,6 +80,122 @@ public Text getKind() {
     return KIND_NAME;
   }
 
+  /** Instead of relying on proto serialization, this
+   *  provides  explicit serialization for OzoneTokenIdentifier.
+   * @return byte[]
+   */
+  public byte[] toUniqueSerializedKey() {
+    ByteBuffer result =
```
Comment (xiaoyuyao) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:88]: Thanks @prashantpogde for working on this. The patch LGTM overall. Have a few questions:
Have you consider leverage the readFields from parent class to handle the basic non-protoc serialization which seems to be much simpler. 

```  
public byte[] toUniqueSerializedKey() throws IOException {
    DataOutputBuffer buf = new DataOutputBuffer();
    super.write(buf);
    WritableUtils.writeEnum(buf, getTokenType());
    // Set s3 specific fields.
    if (getTokenType().equals(S3AUTHINFO)) {
      WritableUtils.writeString(buf, getAwsAccessId());
      WritableUtils.writeString(buf, getSignature());
      WritableUtils.writeString(buf, getStrToSign());
    } else {
      WritableUtils.writeString(buf, getOmCertSerialId());
    }
    return buf.getData();
  }
```

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 10]
```diff
@@ -77,6 +80,122 @@ public Text getKind() {
...
+        ByteBuffer.allocate(4096);
+    result.order(ByteOrder.BIG_ENDIAN);
+    try {
+      result.putLong(getIssueDate());
+      result.putInt(getMasterKeyId());
+      result.putInt(getSequenceNumber());
+
+      result.putLong(getMaxDate());
+
+      result.putInt(getOwner().toString().length());
```
Comment (xiaoyuyao) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:98]: NIT: The Text class is a UTF-8 bytes + length wrapper class. You can use getLength()/getBytes() without toString() conversion. Same apply to owner, realuser and renewer.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 11]
```diff
@@ -77,6 +80,122 @@ public Text getKind() {
...
+    }
+    return result.array();
+  }
+
+  /** Instead of relying on proto deserialization, this
+   *  provides  explicit deserialization for OzoneTokenIdentifier.
+   * @return byte[]
+   */
+  public static OzoneTokenIdentifier fromUniqueSerializedKey(byte[] rawData) {
+    OzoneTokenIdentifier result = newInstance();
```
Comment (xiaoyuyao) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:140]: We could simplify the code by leveraging the readFields from parent class to deserialize non-protoc token id.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java (hunk 9 above)
Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:88]: didn't try this. This looks simpler. Let me try this.

Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:88]: I wanted to do explicit serialization. But this also looks safe serialization.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java (hunk 10 above)
Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:98]: this will change after I use super.write()

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java (hunk 11 above)
Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:140]: Earlier I wanted to do explicit serialization, but your suggestion looks simpler. I will try this.

Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java]: I have uploaded the changes as suggested by @xiaoyuyao Please take a look..

Comment (prashantpogde) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java]: One failure in coverage check looks unreleted.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 12]
```diff
@@ -77,6 +81,57 @@ public Text getKind() {
...
+      super.write(buf);
+      WritableUtils.writeVInt(buf, getTokenType().getNumber());
+      // Set s3 specific fields.
+      if (getTokenType().equals(S3AUTHINFO)) {
+        WritableUtils.writeString(buf, getAwsAccessId());
+        WritableUtils.writeString(buf, getSignature());
+        WritableUtils.writeString(buf, getStrToSign());
+      } else {
+        WritableUtils.writeString(buf, getOmCertSerialId());
+        if (getOmServiceId() != null) {
```
Comment (xiaoyuyao) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:100]: Thanks @prashantpogde  for the update. LGTM overall. Only one question: the null check on line100 should be removed. WritableUtils.writeString will write len=-1 for null and the WritableUtils.readString will populate null back during de-serialization. This way, we will have matched deserialization code on the code in line129-130. 

+1 after that being fixed.

**Comment by xiaoyuyao** [comment]
+1 pending CI.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java (hunk 8 above)
Comment (avijayanhwx) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java]: @ChenSammi Are you OK with resolving this issue?

Comment (avijayanhwx) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java]: This issue is no longer valid in the newer version of the patch. cc @ChenSammi.

**Comment by avijayanhwx** [comment]
Please don't merge this since we are testing out upgrade paths.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java [hunk 13]
```diff
@@ -42,8 +42,9 @@ public OzoneTokenIdentifier fromPersistedFormat(byte[] rawData)
     Preconditions.checkNotNull(rawData,
         "Null byte array can't converted to real object.");
     try {
-      return OzoneTokenIdentifier.readProtoBuf(rawData);
-    } catch (InvalidProtocolBufferException e) {
+      OzoneTokenIdentifier object = OzoneTokenIdentifier.newInstance();
+      return object.fromUniqueSerializedKey(rawData);
+    } catch (BufferUnderflowException e) {
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java:47]: We should catch IOException here and fall back to reading the old format to avoid OM failures to load tokens in old format upon upgrade/restart. We can rewrite the loaded token with the new format like HDDS-3925 upon hitting tokens in old format to avoid future fallbacks. But I'm OK without it given old tokens in old format will be expired in 7 days anyway (unlike the pipeline id) and the new tokens will be written in new format. 

I also suggest we add a unit test like below that can repro the loading failure cases in TestOzoneTokenIdentifier.java. 

  @Test
  public void testTokenPersistence() throws IOException {
    OzoneTokenIdentifier idWrite = getIdentifierInst();
    idWrite.setOmServiceId("defaultServiceId");

    byte[] oldIdBytes = idWrite.getBytes();
    TokenIdentifierCodec idCodec = new TokenIdentifierCodec();

    OzoneTokenIdentifier idRead = null;
    try {
      idRead =  idCodec.fromPersistedFormat(oldIdBytes);
    } catch (IOException ex) {
      Assert.fail("Should not fail to load old token format");
    }
    Assert.assertEquals("Deserialize Serialized Token should equal.",
        idWrite, idRead);
  }
}

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java [hunk 14]
```diff
@@ -42,8 +42,9 @@ public OzoneTokenIdentifier fromPersistedFormat(byte[] rawData)
     Preconditions.checkNotNull(rawData,
         "Null byte array can't converted to real object.");
     try {
-      return OzoneTokenIdentifier.readProtoBuf(rawData);
-    } catch (InvalidProtocolBufferException e) {
+      OzoneTokenIdentifier object = OzoneTokenIdentifier.newInstance();
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java:45]:
```suggestion
      } catch (IOException ex) {
      return OzoneTokenIdentifier.readProtoBuf(rawData);
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java (hunk 13 above)
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java:47]:
```suggestion
    } catch (IOException ex) {
      return OzoneTokenIdentifier.readProtoBuf(rawData);
    } catch (BufferUnderflowException e) {
```

Comment (prashantpogde) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java:56]: done

Comment (prashantpogde) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/codec/TokenIdentifierCodec.java:56]: done

**Comment by prashantpogde** [comment]
Made the changes as suggested by @xiaoyuyao .
